File: packages/api/src/models/Collection.ts
/**
 * Collection Model — User-created bookshelves/groups
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';

interface CollectionAttributes {
  id: string;
  userId: string;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  bookIds: string[];
  createdAt: Date;
  updatedAt: Date;
}

interface CollectionCreationAttributes extends Optional<CollectionAttributes, 'id' | 'description' | 'icon' | 'color' | 'bookIds' | 'createdAt' | 'updatedAt'> {}

export class Collection extends Model<CollectionAttributes, CollectionCreationAttributes> implements CollectionAttributes {
  public id!: string;
  public userId!: string;
  public name!: string;
  public description?: string;
  public icon?: string;
  public color?: string;
  public bookIds!: string[];
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

Collection.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
      onDelete: 'CASCADE',
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        len: [1, 100],
      },
    },
    description: {
      type: DataTypes.STRING,
      allowNull: true,
      validate: {
        len: [0, 500],
      },
    },
    icon: {
      type: DataTypes.STRING,
      allowNull: true,
      defaultValue: 'folder',
    },
    color: {
      type: DataTypes.STRING,
      allowNull: true,
      defaultValue: '#f59e0b',
    },
    bookIds: {
      type: DataTypes.ARRAY(DataTypes.UUID),
      allowNull: false,
      defaultValue: [],
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
    },
  },
  {
    sequelize,
    tableName: 'collections',
    timestamps: true,
    indexes: [
      { fields: ['user_id'] },
      { fields: ['book_ids'], using: 'GIN' },
    ],
  }
);

File: packages/server/app/models/collection.py
"""Collection model."""

import uuid
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Optional
from uuid import UUID

from sqlalchemy import DateTime, ForeignKey, String, Table, Column, text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.db import Base

if TYPE_CHECKING:
    from app.models.book import Book
    from app.models.user import User

# Association table for the many-to-many relationship between collections and books.
collection_books = Table(
    'collection_books',
    Base.metadata,
    Column(
        'collection_id',
        PG_UUID(as_uuid=True),
        ForeignKey('collections.id', ondelete='CASCADE'),
        primary_key=True,
    ),
    Column(
        'book_id',
        PG_UUID(as_uuid=True),
        ForeignKey('books.id', ondelete='CASCADE'),
        primary_key=True,
    ),
    Column(
        'added_at',
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    ),
)


class Collection(Base):
    __tablename__ = 'collections'

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    icon: Mapped[str] = mapped_column(String(255), default='folder')
    color: Mapped[str] = mapped_column(String(255), default='#f59e0b')
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=lambda ctx: datetime.now(tz=timezone.utc),
    )

    user: Mapped['User'] = relationship('User', back_populates='collections')
    books: Mapped[list['Book']] = relationship(
        'Book',
        secondary=collection_books,
        backref='collections',
        lazy='selectin',
    )
